### PR TITLE
Refactor global build info plugin to leverage JavaInstallationRegistry

### LINKED ISSUE #52934: Refactor Java detection logic to use Gradle JavaInstallationRegistry
Gradle 6.2 introduces a [`JavaInstallationRegistry`](https://docs.gradle.org/current/javadoc/org/gradle/jvm/toolchain/JavaInstallationRegistry.html) service which can be used to probe JDK/JRE directories for metadata. We have similar logic in our `GlobalBuildInfoPlugin` to do exactly this as well as deal with caching etc. This could be considerably simplified by using the functionality in Gradle core, which both handles probing JDK locations for information, as well as handling caching.

### PR DESCRIPTION
**Pull request by mark-vieira**
This commit removes the configuration time vs execution time distinction
with regards to certain BuildParms properties. Because of the cost of
determining Java versions for configuration JDK locations we deferred
this until execution time. This had two main downsides. First, we had
to implement all this build logic in tasks, which required a bunch of
additional plumbing and complexity. Second, because some information
wasn't known during configuration time, we had to nest any build logic
that depended on this in awkward callbacks.

We now defer to the JavaInstallationRegistry recently added in Gradle.
This utility uses a much more efficient method for probing Java
installations vs our jrunscript implementation. This, combined with some
optimizations to avoid probing the current JVM as well as deferring
some evaluation via Providers when probing installations for BWC builds
we can maintain effectively the same configuration time performance
while removing a bunch of complexity and runtime cost (snapshotting
inputs for the GenerateGlobalBuildInfoTask was very expensive). The end
result should be a much more responsive build execution in almost all
scenarios.

Closes #52934

### CODE REVIEW

File: buildSrc/src/main/java/org/elasticsearch/gradle/util/Util.java [hunk 1]
```diff
@@ -7,7 +7,7 @@
  * not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
- *    http://www.apache.org/licenses/LICENSE-2.0
+ *         http://www.apache.org/licenses/LICENSE-2.0
```
Comment (rjernst) [inline buildSrc/src/main/java/org/elasticsearch/gradle/util/Util.java]: accidental indentation?

**Review by rjernst — APPROVED** [review]
LGTM

File: buildSrc/src/main/java/org/elasticsearch/gradle/util/Util.java (hunk 1 above)
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/util/Util.java]: Yeah, this is actually messed up in `ide.gradle`. I've fixed that as well.

File: buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java [hunk 2]
```diff
@@ -33,102 +42,168 @@
...
+        } else {
+            testSeed = testSeedProperty;
+        }
+        return testSeed;
+    }
+
+    private static void throwInvalidJavaHomeException(String description, File javaHome, int expectedVersion, int actualVersion) {
+        String message = String.format(
+            Locale.ROOT,
+            "The %s must be set to a JDK installation directory for Java" + " %d but is [%s] corresponding to [%s]",
```
Comment (pugnascotia) [inline buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java]:
```suggestion
            "The %s must be set to a JDK installation directory for Java %d but is [%s] corresponding to [%s]",
```

**Comment by mark-vieira** [comment]
@elasticmachine update branch

**Pull request merged**